pkg/bucket/replication/replication.go

```diff
@@ -0,0 +1,209 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package replication
+
+import (
+	"encoding/xml"
+	"io"
+	"sort"
+	"strings"
+)
+
+// StatusType of Replication for x-amz-replication-status header
+type StatusType string
+
+const (
+	// Pending - replication is pending.
+	Pending StatusType = "PENDING"
+
+	// Complete - replication completed ok.
+	Complete StatusType = "COMPLETE"
+
+	// Failed - replication failed.
+	Failed StatusType = "FAILED"
+
+	// Replica - this is a replica.
+	Replica StatusType = "REPLICA"
+)
+
+// String returns string representation of status
+func (s StatusType) String() string {
+	return string(s)
+}
+
+var (
+	errReplicationTooManyRules        = Errorf("Replication configuration allows a maximum of 1000 rules")
+	errReplicationNoRule              = Errorf("Replication configuration should have at least one rule")
+	errReplicationOverlappingPrefix   = Errorf("Replication configuration has rules with overlapping prefix")
+	errReplicationDestinationMismatch = Errorf("The destination bucket must be same for all rules")
+)
+
+// Config - object lock configuration specified in
+// https://docs.aws.amazon.com/AmazonS3/latest/API/Type_API_ObjectLockConfiguration.html
+type Config struct {
+	XMLName xml.Name `xml:"ReplicationConfiguration"`
+	Rules   []Rule   `xml:"Rule"`
+}
+
+// Maximum 2MiB size per replication config.
+const maxReplicationConfigSize = 1 << 21
+
+// ParseConfig parses ReplicationConfiguration from xml
+func ParseConfig(reader io.Reader) (*Config, error) {
+	config := Config{}
+	if err := xml.NewDecoder(io.LimitReader(reader, maxReplicationConfigSize)).Decode(&config); err != nil {
+		return nil, err
+	}
+	return &config, nil
+}
+
+// Validate - validates the replication configuration
+func (c Config) Validate(bucket string) error {
+	// replication config can't have more than 1000 rules
+	if len(c.Rules) > 1000 {
+		return errReplicationTooManyRules
+	}
+	// replication config should have at least one rule
+	if len(c.Rules) == 0 {
+		return errReplicationNoRule
+	}
+	// Validate all the rules in the replication config
+	targetMap := make(map[string]string)
+	for _, r := range c.Rules {
+		if len(targetMap) == 0 {
+			targetMap[r.Destination.Bucket] = ""
+		}
+		if _, ok := targetMap[r.Destination.Bucket]; !ok {
+			return errReplicationDestinationMismatch
+		}
+		if err := r.Validate(bucket); err != nil {
+			return err
+		}
+	}
+	// Compare every rule's prefix with every other rule's prefix
```

Comment:
Where is this limitation documented? I couldn't find any references to it anywhere.

Reply:
I see how this could potentially lead to conflicting rules. But in the lifecycle I think the first matches rule is applied and overlapping is allowed. Could it the same here?

Reply:
I think its here? https://docs.aws.amazon.com/AmazonS3/latest/user-guide/enable-replication.html

Reply:
> If the bucket has existing replication rules, you are instructed to set a priority for the rule. You must set a priority for the rule to avoid conflicts caused by objects that are included in the scope of more than one rule. In the case of overlapping rules, Amazon S3 uses the rule priority to determine which rule to apply. The higher the number, the higher the priority. For more information about rule priority, see Replication Configuration Overview in the Amazon Simple Storage Service Developer Guide.

https://docs.aws.amazon.com/AmazonS3/latest/dev/replication-add-config.html

Reply:
We should only be supporting XMLv2

Reply:
removed the overlapping rules check, v2 looks only at priority ordering.